### ISSUE TITLE
Setting controller.runAsUser != 1000 leads to failing init container

### ISSUE THREAD

**Issue by schnatterer** [jenkinsci/helm-charts#398]
**Describe the bug**
When setting `controller.runAsUser` to a value other than the built in jenkins user id 1000 the `init` container fails.

**Version of Helm and Kubernetes**:

Helm Version:

```console
$ helm version
version.BuildInfo{Version:"v3.6.1", GitCommit:"61d8e8c4a6f95540c15c6a65f36a6dd0a45e7a2f", GitTreeState:"clean", GoVersion:"go1.16.5"}
```

Kubernetes Version:

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:11:29Z", GoVersion:"go1.16.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.3+k3s3", GitCommit:"0e4fbfefe1dd8734756dfa4f9ab4fc89665cece4", GitTreeState:"clean", BuildDate:"2020-11-13T07:19:02Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Which version of the chart**:
3.4.0

**What happened**:
Started the chart with `controller.runAsUser=1001`  results in `jenkins-0   0/2     Init:CrashLoopBackOff`

The log of `init`containers shows 

```
 java.nio.file.NoSuchFileException: ?/.cache      
```

Presumably the errors is caused here:

https://github.com/jenkinsci/helm-charts/blob/6e829d121c81cc935978a7a436cf0d93f53fa3e2/charts/jenkins/templates/config.yaml#L38

Because the `jenkins-plugin-cli` uses the system var `user.home` [as base dir](
https://github.com/jenkinsci/plugin-installation-manager-tool/blob/13693591e4cb9415a4e87868045fb113a0b9307c/plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/config/Settings.java#L24). I presume that the `user.home` home is initialized from `/etc/passwd`. The default UID 1000 is contained in `/etc/passwd` so it works. For other UIDs (e.g. 1001) there is no entry in `/etc/passwd` so `?` is returned. 
Explicitly setting the `HOME` variable does not solve this issue. 
Adding `--Duser.home="{{ .Values.controller.jenkinsHome }}` to the `jenkins-plugin-cli` call might do it, though.

**What you expected to happen**:
The pod starts successfully.

**How to reproduce it** (as minimally and precisely as possible):
```
helm upgrade -i jenkins --set controller.runAsUser=1001 --version=3.4.0 jenkins/jenkins
```

**Anything else we need to know**:
I'm a bit time-constrained right now so I can't validate the fix proposed above. Still wanted to document it.
I might create a PR at a later stage.

**Comment by torstenwalter** [comment]
Do you think there is something the helm chart could do to fix this? It sounds more like an issue of the docker image. Some people might use customized images where a different user id is present. So allowing this configuration makes sense.

**Comment by schnatterer** [comment]
If a custom image is the reason for this option - then a warning in the comments of `controller.runAsUser` in `values.yaml` about changing the UID when running the default image could prevent running into this issue.

A fix that would have worked for me would be to set `-Duser.home="{{ .Values.controller.jenkinsHome }}` here:

https://github.com/jenkinsci/helm-charts/blob/0afd66178a112cee7dd8579ea79ace51db42dbe5/charts/jenkins/templates/config.yaml#L37-L38

But I can't judge if this would cause collateral damage. I also don't know if this issue would occur in the `else` branch  
https://github.com/jenkinsci/helm-charts/blob/0afd66178a112cee7dd8579ea79ace51db42dbe5/charts/jenkins/templates/config.yaml#L39-L40
and if something like `HOME={{ .Values.controller.jenkinsHome }}  /usr/local/bin/install-plugins.sh ...` would fix it.

**Comment by torstenwalter** [comment]
> A fix that would have worked for me would be to set -Duser.home="{{ .Values.controller.jenkinsHome }}

We could set that by default. No matter which runAsUser is specified.

Do you want to create a PR for that?

**Comment by torstenwalter** [comment]
@timja do you know if setting home would be respected by the CLI?

**Comment by timja** [comment]
I would walk back a bit,

Why are you doing this? You may have other issues if you continue down this route with the persistent volume as well.

If you want to change the user you will need to re-build the jenkins upstream image yourself and set the argument for it. https://github.com/jenkinsci/docker/blob/master/11/debian/buster/hotspot/Dockerfile#L23

**Comment by schnatterer** [comment]
@timja 
We used to set the UID to the local user's UID in a project (cloudogu/gitops-playground@111e666f6d6f908a8c8e7ffe9c2c3ae1f1ee431d) that runs on a local cluster in order to avoid file permissions issues on bind mounted folders. 
We had no other issues with files inside the images. But I can see that those might turn up for other users.

We ended up removing the option and it works for us.
The behaviour was still somewhat confusing to me, so I thought it was worth the issue.

So the only takeaway here would be a warning in the comments above the option stating that changing the UID might cause issues if the underlying image does not provide an entry in `/etc/passwd` and appropriate file permissions?

@torstenwalter I wanted to discuss this first, in order to see if a PR is relevant at all.

**Comment by timja** [comment]
> So the only takeaway here would be a warning in the comments above the option stating that changing the UID might cause issues if the underlying image does not provide an entry in /etc/passwd and appropriate file permissions?

I think so. If you check the https://github.com/jenkinsi/docker repo there's quite a few issues related to this, but it mostly comes down to docs.

**Comment by vrmerlin** [comment]
Are there any work arounds for this problem? I'm trying to deploy this Helm chart to our cluster, using a specific UID and GID, and it's failing on startup as described above. Far as I can tell, there is no way to run this chart as a different user. I'm really stuck on this -- any suggestions??

**Comment by timja** [comment]
rebuild the container image with the uid / gid that you want, you can adjust the ARGS at build time:
https://github.com/jenkinsci/docker/blob/3c143f2afc2563686341225b106f9c695574f024/17/debian/bullseye/hotspot/Dockerfile#L41-L44

**Comment by vrmerlin** [comment]
> rebuild the container image with the uid / gid that you want, you can adjust the ARGS at build time: https://github.com/jenkinsci/docker/blob/3c143f2afc2563686341225b106f9c695574f024/17/debian/bullseye/hotspot/Dockerfile#L41-L44

I don't believe that's the problem. I know it's possible to override the UID and GID for the container, because some files are being created with the proper permissions. But then it fails on the .cache directory creation. As described above, this is because the user.home is not being passed into the Java process.

I've tried three different ways, but none seem to be working so far

```
controller:
  javaOpts: "-Duser.home=/var/jenkins_home"
  initContainerEnv:
    - name: JAVA_HOME
      value: "/var/jenkins_home"   
  containerEnv:
    - name: JAVA_HOME
      value: "/var/jenkins_home"  
```
